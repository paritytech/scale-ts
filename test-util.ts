/// <reference lib="deno.unstable"/>

import { assertEquals, assertThrows } from "https://deno.land/std@0.161.0/testing/asserts.ts"
import { assertSnapshot } from "https://deno.land/std@0.161.0/testing/snapshot.ts"
import { AnyShape, assert, DecodeBuffer, Shape, ShapeAssertError } from "./common/mod.ts"

const [lipsum, words, cargoLock] = ["lipsum.txt", "words.txt", "Cargo.lock"].map((fileName) => () =>
  Deno.readTextFile(fileName)
)
export const files = { lipsum: lipsum!, words: words!, cargoLock: cargoLock! }

export function testShape<T>(
  shape: Shape<T>,
  values: NoInfer<T>[] | Record<string, NoInfer<T> | (() => NoInfer<T>)>,
  async?: boolean,
): void
export function testShape<T>(
  shape: Shape<T>,
  values: T[] | Record<string, T | (() => T)>,
  async = false,
) {
  for (const key in values) {
    let value = values[key as never] as T | (() => T)
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key
    Deno.test(`${Deno.inspect(shape)} ${label}`, async (t) => {
      if (typeof value === "function") {
        value = (value as () => T)()
      }
      assert(shape, value)
      const encoded = async ? await shape.encodeAsync(value) : shape.encode(value)
      await assertSnapshot(t, encoded, { serializer: serializeU8A })
      const decodeBuffer = new DecodeBuffer(encoded)
      const decoded = shape.subDecode(decodeBuffer)
      assertEquals(decoded, value)
      assertEquals(decodeBuffer.index, encoded.length)
      assert(shape, decoded)
    })
  }
}

export function testInvalid(shape: AnyShape, values: unknown[] | Record<string, unknown | (() => unknown)>) {
  for (const key in values) {
    let value: unknown = values[key as never]
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key
    Deno.test(`${Deno.inspect(shape)} invalid ${label}`, async (t) => {
      if (typeof value === "function") {
        value = (value as () => unknown)()
      }
      await assertThrowsSnapshot(t, () => assert(shape as Shape<any>, value), ShapeAssertError)
    })
  }
}

function serializeU8A(array: Uint8Array) {
  return [...array].map((x) => x.toString(16).padStart(2, "0")).join("\n")
}

type NoInfer<T> = T extends infer U ? U : never
export function benchShape<T>(name: string, shape: Shape<T>, value: NoInfer<T>): void
export function benchShape<T>(name: string, shape: Shape<T>, value: T) {
  const encoded = shape.encode(value)
  Deno.bench(`- ${name} (${encoded.length}B) [encode] `, () => {
    shape.encode(value)
  })
  Deno.bench(`  ${name} (${encoded.length}B) [decode]`, () => {
    shape.decode(encoded)
  })
}

export async function assertThrowsSnapshot(
  t: Deno.TestContext,
  fn: () => unknown,
  ctor: abstract new(...args: any) => Error = Error,
) {
  try {
    fn()
  } catch (error) {
    if (!(error instanceof ctor)) {
      throw new Error("Expected " + ctor.name)
    }
    return await assertSnapshot(t, `${error.name}: ${error.message}`, { serializer: (x) => x })
  }
  throw new Error("Expected function to throw")
}

export { assertEquals, assertSnapshot, assertThrows }
