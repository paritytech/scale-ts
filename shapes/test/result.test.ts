import * as $ from "../../mod.ts"
import { assertThrows, testInvalid, testShape } from "../../test-util.ts"

class StrErr extends Error {
  constructor(readonly str: string) {
    super()
    this.stack = this.message = "StrErr: " + this.str
  }
}

const $strError = $.instance(StrErr, $.tuple($.str), (err: StrErr) => [err.str])

testShape($.result($.str, $strError), [
  "ok",
  new StrErr("err"),
])

testInvalid($.result($.str, $strError), [
  null,
  undefined,
  Object.assign(new Error("foo"), { stack: "Error: foo" }),
  new StrErr(null!),
])

Deno.test("result roundtrip error", async () => {
  assertThrows(() => $.result($.result($.u8, $strError), $strError))
  assertThrows(() => $.result($.withMetadata($.metadata("$foo"), $.result($.u8, $strError)), $strError))
  assertThrows(() => {
    const $foo = $.result($.u8, $strError)
    $foo.metadata = []
    $.result($foo, $strError)
      // Ok(Err(""))
      .decode(new Uint8Array([0, 1, 0]))
  })
})
