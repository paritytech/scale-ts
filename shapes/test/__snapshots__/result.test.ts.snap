export const snapshot = {};

snapshot[`\$.result(
  \$.str,
  \$.instance(
    [class StrErr extends Error],
    \$.tuple(\$.str),
    [Function (anonymous)]
  )
) "ok" 1`] = `
00
08
6f
6b
`;

snapshot[`\$.result(
  \$.str,
  \$.instance(
    [class StrErr extends Error],
    \$.tuple(\$.str),
    [Function (anonymous)]
  )
) { [StrErr: err] str: "err", message: "StrErr: err" } 1`] = `
01
0c
65
72
72
`;

snapshot[`\$.result(
  \$.str,
  \$.instance(
    [class StrErr extends Error],
    \$.tuple(\$.str),
    [Function (anonymous)]
  )
) invalid null 1`] = `ScaleAssertError: typeof value !== "string"`;

snapshot[`\$.result(
  \$.str,
  \$.instance(
    [class StrErr extends Error],
    \$.tuple(\$.str),
    [Function (anonymous)]
  )
) invalid undefined 1`] = `ScaleAssertError: typeof value !== "string"`;

snapshot[`\$.result(
  \$.str,
  \$.instance(
    [class StrErr extends Error],
    \$.tuple(\$.str),
    [Function (anonymous)]
  )
) invalid [Error: foo] 1`] = `ScaleAssertError: !(value instanceof StrErr)`;

snapshot[`\$.result(
  \$.str,
  \$.instance(
    [class StrErr extends Error],
    \$.tuple(\$.str),
    [Function (anonymous)]
  )
) invalid { [StrErr: null] str: null, message: "StrErr: null" } 1`] = `ScaleAssertError: typeof value#arguments[0] !== "string"`;
